native/src/main/java/io/ballerina/lib/data/jsondata/json/JsonTraverse.java

```diff
@@ -117,6 +118,16 @@ private Object traverseJson(Object json, Type type) {
                     RecordType recordType = (RecordType) referredType;
                     fieldHierarchy.push(JsonCreator.getAllFieldsInRecord(recordType));
                     restType.push(recordType.getRestFieldType());
+                    if (recordType.isReadOnly()) {
+                        Object value = traverseMapJsonOrArrayJson(json,
+                                        ValueCreator.createMapValue(TypeCreator
+                                                .createMapType(PredefinedTypes.TYPE_ANYDATA)), referredType);
```

Comment:
Can't this be `map<json>` instead? 

Btw this isn't optimal anyway, right? Creating two values? I see that this is what we are doing for readonly intersections also, but could still be a problem with large values. Ideally, we should use the APIs that accept the initial values to create an immutable value. I believe we discussed this initially also, but was complicated to implement maybe. Something to consider sometime later.